File: 7_TheTimeInWords/solution2.py
#!/bin/python3
# azusaw

import math
import os
import random
import re
import sys

#
# Complete the 'timeInWords' function below.
#
# The function is expected to return a STRING.
# The function accepts following parameters:
#  1. INTEGER h
#  2. INTEGER m
#

def timeInWords(h, m):
    s = []
    if m == 0:
        s = [convertNumToStr(h), "o' clock"]
    elif m % 15 == 0:
        s.append("half" if m == 30 else "quarter") 
        s.append("to" if m == 45 else "past") 
        s.append(convertNumToStr(h + 1 if m == 45 else h))
    else:
        s.append(convertNumToStr(m if m < 30 else 60 - m))
        s.append("minute" if m == 1 or m == 59 else "minutes")
        s.append("past" if m < 30 else "to")
        s.append(convertNumToStr(h if m < 30 else h + 1))
        
    return " ".join(s)

def convertNumToStr(n):
    numl = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine", "ten", "eleven", "twelve"]
    prefix = ["thir", numl[3], "fif", numl[5], numl[6], "eigh", numl[8]]
    if n < 13:
        return numl[n - 1]
    elif n < 20:
        return prefix[n - 13] + "teen"
    elif n == 20:
        return "twenty"
    return "twenty " + numl[n - 21]
        
if __name__ == '__main__':
    fptr = open(os.environ['OUTPUT_PATH'], 'w')

    h = int(input().strip())

    m = int(input().strip())

    result = timeInWords(h, m)

    fptr.write(result + '\n')

    fptr.close()

File: 7_TheTimeInWords/solution2.cpp
#include <bits/stdc++.h>

using namespace std;

string ltrim(const string &);
string rtrim(const string &);

/*
 * Complete the 'timeInWords' function below.
 *
 * The function is expected to return a STRING.
 * The function accepts following parameters:
 *  1. INTEGER h
 *  2. INTEGER m
 */

string numToWords(int n) {
    
    string num[21]={
                    "one",
                    "two",
                    "three",
                    "four",
                    "five",
                    "six",
                    "seven",
                    "eight",
                    "nine",
                    "ten",
                    "eleven",
                    "twelve",
                    "thirteen",
                    "fourteen",
                    "quarter",
                    "sixteen",
                    "seventeen",
                    "eighteen",
                    "nineteen",
                    "twenty",
                    "half"
                    };

    if (n<=20) return num[n-1];
    else if (n==30) return num[20];
    else {
        return num[19] + " " + num[n-21];
    }
}

string timeInWords(int h, int m) {
    string res;
    
    if (m==0) {
        return numToWords(h) + " o' clock";
    }
    else if (m==15||m==30) {
        return numToWords(m) + " past " + numToWords(h);
    }
    else if (m==1) {
        return numToWords(m) + " minute past " + numToWords(h);
    }
    if (m<=30) {
        return numToWords(m) + " minutes past " + numToWords(h);
    }
    if (m==45) {
        return numToWords(60-m) + " to " + numToWords(h+1);
    }
    if (m==59) {
            return numToWords(60-m) + " minute to " + numToWords(h+1);
    }
    else {
            return numToWords(60-m) + " minutes to " + numToWords(h+1);
    }   
}

int main()
{
    ofstream fout(getenv("OUTPUT_PATH"));

    string h_temp;
    getline(cin, h_temp);

    int h = stoi(ltrim(rtrim(h_temp)));

    string m_temp;
    getline(cin, m_temp);

    int m = stoi(ltrim(rtrim(m_temp)));

    string result = timeInWords(h, m);

    fout << result << "\n";

    fout.close();

    return 0;
}

string ltrim(const string &str) {
    string s(str);

    s.erase(
        s.begin(),
        find_if(s.begin(), s.end(), not1(ptr_fun<int, int>(isspace)))
    );

    return s;
}

string rtrim(const string &str) {
    string s(str);

    s.erase(
        find_if(s.rbegin(), s.rend(), not1(ptr_fun<int, int>(isspace))).base(),
        s.end()
    );

    return s;
}
